[InstCombine] Pattern match minmax calls for unsigned saturation.

huihzhang opened the pull request:
Extend the facility from D68651 to match the following patterns for unsigned saturation:
1. fold smax(0, sub(zext(A), zext(B))) into usub_sat;
2. fold umin(UINT_MAX, add(zext(A), zext(B))) into uadd_sat.

Proofs:
uadd_sat: https://alive2.llvm.org/ce/z/v-LJZr
usub_sat: https://alive2.llvm.org/ce/z/Veoypa

huihzhang committed:
[InstCombine] Add test for unsigned addsub saturation. NFC.

huihzhang committed:
[InstCombine] Pattern match minmax calls for unsigned saturation.

This patch matches the following patterns for unsigned saturation:
1) fold smax(UINT_MIN, smin(UINT_MAX, sub(zext(A), zext(B)))) into usub_sat,
   where smin smax could be reversed.
2) fold umin(UINT_MAX, add(zext(A), zext(B))) into uadd_sat.

Note that this patch extends the signed saturation (sadd|ssub_sat) pattern
matching from D68651.

goldsteinn commented:
Hi,

can you please add alive2 proofs? See: https://llvm.org/docs/InstCombineContributorGuide.html#proofs for more details.

huihzhang committed:
Fix pattern match for umin(UINT_MAX, BinOp(zext(A), zext(B))).

Bail out if BinOp is not known non-negative.

huihzhang commented:
Thanks @goldsteinn!
Caught a bug while doing alive2 testing. Update fix in [commit#3](https://github.com/llvm/llvm-project/pull/99250/commits/603655d2ed39e4622756924ae4d449210bc553fd) .

When trying to fold "umin(UINT_MAX, BinOp(zext(A), zext(B)))" into uadd_sat, we need to make sure BinOp is known non-negative.

Please also see alive2 verification below:
uadd_sat: https://alive2.llvm.org/ce/z/v-LJZr
usub_sat: https://alive2.llvm.org/ce/z/h34acm

dtcxzyw commented on llvm/lib/Transforms/InstCombine/InstCombineCalls.cpp:
```diff
@@ -1117,68 +1117,92 @@ static Instruction *moveAddAfterMinMax(IntrinsicInst *II,
+  //    smax(SINT_MIN, smin(SINT_MAX, add|sub(sext(A), sext(B))))
+  //    Where the smin and smax could be reversed.
+  // 2. A tree of unsigned saturation:
+  //    smax(UINT_MIN, smin(UINT_MAX, sub(zext(A), zext(B))))
```
`UINT_MAX` is -1. It looks incorrect.

goldsteinn commented:
The `smin(UINT_MAX, (sub (zext A), (zext B))` is dead code: https://alive2.llvm.org/ce/z/AZ-puW

huihzhang commented:
Thanks @dtcxzyw @goldsteinn for the feedbacks! 
I will post an update, and simplify into below pattern match rules:
1. fold smax(UINT_MIN, sub(zext(A), zext(B))) into usub_sat,
2. fold umin(UINT_MAX, add(zext(A), zext(B))) into uadd_sat.

dtcxzyw commented:
@huihzhang Please update the alive2 link for `smax(UINT_MIN, sub(zext(A), zext(B))) -> usub_sat` in the PR description.

dtcxzyw commented on llvm/test/Transforms/InstCombine/uaddsub_sat.ll:
```diff
@@ -0,0 +1,517 @@
+; NOTE: Assertions have been autogenerated by utils/update_test_checks.py UTC_ARGS: --version 5
+; RUN: opt < %s -passes=instcombine -S | FileCheck %s
+
+target datalayout = "e-m:e-p:32:32-Fi8-i64:64-v128:64:128-a:0:32-n32-S64"
```
```suggestion
```

dtcxzyw commented on llvm/test/Transforms/InstCombine/uaddsub_sat.ll:
```diff
@@ -0,0 +1,517 @@
+  %conv = zext i32 %a to i64
+  %conv1 = zext i32 %b to i64
+  %add = add i64 %conv1, %conv
+  %0 = icmp ult i64 %add, 4294967295
```
Use named values.

dtcxzyw commented on llvm/test/Transforms/InstCombine/uaddsub_sat.ll:
```diff
@@ -0,0 +1,517 @@
+}
+
+declare void @use64(i64)
+declare i64 @llvm.umin.i64(i64, i64)
+declare i64 @llvm.smin.i64(i64, i64)
+declare i64 @llvm.smax.i64(i64, i64)
+declare i32 @llvm.umin.i32(i32, i32)
+declare i32 @llvm.smin.i32(i32, i32)
+declare i32 @llvm.smax.i32(i32, i32)
+declare <8 x i32> @llvm.umin.v8i32(<8 x i32>, <8 x i32>)
+declare <2 x i128> @llvm.smin.v2i128(<2 x i128>, <2 x i128>)
+declare <2 x i128> @llvm.smax.v2i128(<2 x i128>, <2 x i128>)
```
```suggestion
```

Intrinsic declarations will be implied by LLParser.

huihzhang committed:
Simplify unsigned saturation pattern match rules:

1. fold smax(UINT_MIN, sub(zext(A), zext(B))) into usub_sat;
2. fold umin(UINT_MAX, add(zext(A), zext(B))) into uadd_sat.

huihzhang commented:
@dtcxzyw @goldsteinn, I have pushed an update to use the simplified pattern match rules for unsigned saturation.
Please help take a look, thanks a lot!

dtcxzyw commented on llvm/lib/Transforms/InstCombine/InstCombineCalls.cpp:
```diff
@@ -1117,68 +1117,108 @@ static Instruction *moveAddAfterMinMax(IntrinsicInst *II,
+  // Pattern match for unsigned saturation.
+  if (match(&MinMax1, m_UMin(m_BinOp(AddSub), m_APInt(MaxValue)))) {
+    // Bail out if AddSub could be negative.
+    if (!isKnownNonNegative(AddSub, SQ.getWithInstruction(AddSub)))
```
Is this check necessary? I don't see any `llvm.assume` in the alive2 proof.

huihzhang replied:
This check was used to reject "umin(UINT_MAX, sub) -> usub_sat". 
I pushed a new update to check for BinOp opcode 'Add', so that !isKnownNonNegative check can be deleted.

dtcxzyw commented on llvm/lib/Transforms/InstCombine/InstCombineCalls.cpp:
```diff
@@ -1117,68 +1117,108 @@ static Instruction *moveAddAfterMinMax(IntrinsicInst *II,
-                   m_SMax(m_Instruction(MinMax2), m_APInt(MinValue)))) {
-    if (!match(MinMax2, m_SMin(m_BinOp(AddSub), m_APInt(MaxValue))))
+    IsUnsignedSaturate = true;
+  } else if (match(&MinMax1, m_SMax(m_BinOp(AddSub), m_APInt(MinValue)))) {
```
```suggestion
  } else if (match(&MinMax1, m_SMax(m_BinOp(AddSub), m_Zero()))) {
```

dtcxzyw commented on llvm/lib/Transforms/InstCombine/InstCombineCalls.cpp:
```diff
@@ -1117,68 +1117,108 @@ static Instruction *moveAddAfterMinMax(IntrinsicInst *II,
   Value *BT = Builder.CreateTrunc(AddSub->getOperand(1), NewTy);
   Value *Sat = Builder.CreateCall(F, {AT, BT});
-  return CastInst::Create(Instruction::SExt, Sat, Ty);
+  return CastInst::Create(
+      IsUnsignedSaturate ? Instruction::ZExt : Instruction::SExt, Sat, Ty);
```
```suggestion
  return CastInst::CreateIntegerCast(Sat, Ty, /*IsSigned=*/!IsUnsignedSaturate);
```

goldsteinn commented on llvm/lib/Transforms/InstCombine/InstCombineCalls.cpp:
```diff
@@ -1117,68 +1117,108 @@ static Instruction *moveAddAfterMinMax(IntrinsicInst *II,
+  //    Where the smin and smax could be reversed.
+  // 2. A tree of unsigned saturation:
+  //    smax(UINT_MIN, sub(zext(A), zext(B)))
+  //    Or umin(UINT_MAX, add(zext(A), zext(B))).
```
Can you update the comment as to which variant of `[s|u][add|sub]_sat` the different min/max patterns correspond to?

goldsteinn commented on llvm/lib/Transforms/InstCombine/InstCombineCalls.cpp:
```diff
@@ -1117,68 +1117,108 @@ static Instruction *moveAddAfterMinMax(IntrinsicInst *II,
+                         : ComputeMaxSignificantBits(Op1, 0, AddSub);
+  unsigned NewBitWidth = IsUnsignedSaturate
+                             ? std::max(Op0MaxBitWidth, Op1MaxBitWidth)
+                             : (*MaxValue + 1).logBase2() + 1;
```
The compute known bits stuff should be after the:
```
  if (!shouldChangeType(Ty->getScalarType()->getIntegerBitWidth(), NewBitWidth))
    return nullptr;

  // Also make sure that the inner min/max and the add/sub have one use.
```

Checks below.

huihzhang replied:
The problem is when trying to fold "smax(UINT_MIN, sub(zext(A), zext(B))) -> usub_sat", the MaxValue is not given. I was trying to use computeKnownBits to determine NewBitWidth.
I pushed a new update to first try setting NewBitWidth to half of the bitwidth of MinMax1, when MaxValue is not given.
Later use the results from computeKnownBits to try to reduce NewBitWidth further, and check for legality.
Please let me know if this approach is more sensible?

goldsteinn commented on llvm/lib/Transforms/InstCombine/InstCombineCalls.cpp:
```diff
@@ -1117,68 +1117,108 @@ static Instruction *moveAddAfterMinMax(IntrinsicInst *II,
   if (AddSub->getOpcode() == Instruction::Add)
-    IntrinsicID = Intrinsic::sadd_sat;
+    IntrinsicID =
+        IsUnsignedSaturate ? Intrinsic::uadd_sat : Intrinsic::sadd_sat;
```
I don't see any logic thats checking the right binop for the given minmax pattern. I.e `umin(UINT_MAX, add)` -> `uadd_sat` is okay, but `umin(UINT_MAX, sub)` -> `usub_sat` isn.t

huihzhang replied:
"umin(UINT_MAX, sub) -> usub_sat" was previously rejected by "!isKnownNonNegative".
`
-  if (match(&MinMax1, m_UMin(m_BinOp(AddSub), m_APInt(MaxValue)))) {
-    // Bail out if AddSub could be negative.
-    if (!isKnownNonNegative(AddSub, SQ.getWithInstruction(AddSub)))
-      return nullptr;
+  if (match(&MinMax1, m_UMin(m_BinOp(AddSub), m_APInt(MaxValue))) &&
+      AddSub->getOpcode() == Instruction::Add) {
`
In the new update, I check for BinOp opcode equals 'Add', to make sure we don't accept umin(UINT_MAX, sub) case.

huihzhang committed:
Address review comments.

huihzhang commented:
Thanks a lot @dtcxzyw @goldsteinn for the review feedbacks!
I pushed an update to address review comments.
My main concern is when folding "smax(UINT_MIN, sub(zext(A), zext(B)))" into usub_sat, the MaxValue is not given.
I am using a different approach to first estimate NewBitWith, later use the results from computeKnownBits to try to reduce NewBitWidth , and check for legality. 
Please help take a look, and let me know if this approach is more sensible?
